### PR TITLE
feat(DTFS2-7786): abandon amendment maker checker email

### PR DESCRIPTION
**Pull request by rkontogianni**
# Pull Request

## Introduction :pencil2:

This PR is about to be sent an email to maker and checker once the the maker abandon an amendment. 

## Resolution :heavy_check_mark:
Update post abandon portal facility amendment controllers with email variables
Update post cancel portal facility amendment controllers with empty email variables
Add mapAbandonEmailVariable to get the email variables for a GovNotify email when maker abandon the amendment
Add validate-delete-portal-facility-amendment-payload in portal-api
Update validate-delete-portal-facility-amendment-payload in dtfs-central-api
Add unit tests

## Miscellaneous :heavy_plus_sign:

List any additional fixes or improvements.

## Screenshot :camera_flash:
image
Screenshot 2025-06-24 132721

### CODE REVIEW

File: dtfs-central-api/src/v1/controllers/portal/facility/delete-amendment.controller.ts [hunk 1]
```diff
@@ -16,19 +18,34 @@ export type DeletePortalAmendmentRequest = CustomExpressRequest<{ params: Delete
...
+      throw new AmendmentNotFoundError(amendmentId, facilityId);
+    }
+
     await TfmFacilitiesRepo.deletePortalFacilityAmendment({
       facilityId: facilityMongoId,
       amendmentId: amendmentMongoId,
       auditDetails,
     });
 
+    const amendmentStatus = amendment.status;
```
Comment (zainUKEF) [inline dtfs-central-api/src/v1/controllers/portal/facility/delete-amendment.controller.ts:41]:
```suggestion
    const { status }= amendment;
```

File: dtfs-central-api/src/v1/controllers/portal/facility/delete-amendment.controller.ts [hunk 2]
```diff
@@ -16,19 +18,34 @@ export type DeletePortalAmendmentRequest = CustomExpressRequest<{ params: Delete
...
+    if (amendmentStatus === PORTAL_AMENDMENT_STATUS.FURTHER_MAKERS_INPUT_REQUIRED) {
```
Comment (zainUKEF) [inline dtfs-central-api/src/v1/controllers/portal/facility/delete-amendment.controller.ts:42]:
```suggestion
    if (status === PORTAL_AMENDMENT_STATUS.FURTHER_MAKERS_INPUT_REQUIRED) {
```

File: dtfs-central-api/src/v1/controllers/portal/facility/delete-amendment.controller.test.ts [hunk 3]
```diff
@@ -1,32 +1,43 @@
...
 import { TfmFacilitiesRepo } from '../../../../repositories/tfm-facilities-repo';
+import externalApi from '../../../../external-api/api';
 
 const facilityId = new ObjectId().toString();
 const amendmentId = new ObjectId().toString();
 
 const mockDeletePortalFacilityAmendment = jest.fn();
+const mockfindOneAmendmentByFacilityIdAndAmendmentId = jest.fn();
+
+const mockAmendment = { facilityId, type: AMENDMENT_TYPES.PORTAL, status: PORTAL_AMENDMENT_STATUS.FURTHER_MAKERS_INPUT_REQUIRED };
```
Comment (zainUKEF) [inline dtfs-central-api/src/v1/controllers/portal/facility/delete-amendment.controller.test.ts:17]: should we not have the full amendment structure here to be consistent? can use aPortalFacilityAmendment from libs/common

File: gef-ui/server/controllers/amendments/helpers/map-abandon-amendment-email-variables.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  facility,
+  amendment,
+  user,
+  checker,
+}: {
+  deal: Deal | null;
+  facility: Facility | null;
+  amendment: PortalFacilityAmendmentWithUkefId | null;
+  user: PortalSessionUser | null;
+  checker: PortalSessionUser | null;
```
Comment (zainUKEF) [inline gef-ui/server/controllers/amendments/helpers/map-abandon-amendment-email-variables.ts:29]: do these have to be null or could they be undefined? if undefined, could do deal?:....

Comment (rkontogianni) [reply gef-ui/server/controllers/amendments/helpers/map-abandon-amendment-email-variables.ts:29]: @zainUKEF I added these, because I wanted to add a unit test 'when deal, facility, amendment user and checker are null'

File: portal-api/src/v1/gef/routes.js [hunk 5]
```diff
@@ -139,7 +140,7 @@ router
   .all(validatePortalFacilityAmendmentsEnabled, mongoIdValidation('facilityId'), mongoIdValidation('amendmentId'), handleExpressValidatorResult)
   .get(validateUserHasAtLeastOneAllowedRole({ allowedRoles: [MAKER, CHECKER] }), getAmendment)
   .patch(validateUserHasAtLeastOneAllowedRole({ allowedRoles: [MAKER] }), validatePatchPortalFacilityAmendmentPayload, patchAmendment)
-  .delete(validateUserHasAtLeastOneAllowedRole({ allowedRoles: [MAKER] }), deleteAmendment);
+  .delete(validateUserHasAtLeastOneAllowedRole({ allowedRoles: [MAKER] }), validateDeletePortalFacilityAmendmentPayload, deleteAmendment);
```
[the remark this answers is not in the document]
Comment (rkontogianni) [reply portal-api/src/v1/gef/routes.js:143]: @abhi-markan there are in the .all these mongo id validators

File: gef-ui/server/controllers/amendments/helpers/map-abandon-amendment-email-variables.ts (hunk 4 above)
Comment (abhi-markan) [reply gef-ui/server/controllers/amendments/helpers/map-abandon-amendment-email-variables.ts:29]: Apologies @rkontogianni for the confusion, if we are using types then no need to add `null` type specifically to cater for a unit test case, in that case those test cases are void due to strict typescript compilation check.

Can the null please be removed if there is no real word scenario where these parameters can ever be null.

I hope this makes sense?

File: gef-ui/server/controllers/amendments/helpers/map-abandon-amendment-email-variables.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+}) => {
+  // Check if any of the required parameters are null
+  if (!deal || !facility || !amendment || !user || !checker) {
+    throw new Error('Deal, Facility, Amendment, User or Checker are null');
+  }
+
+  const commonVariables = mapCommonEmailVariables({ deal, facility, amendment });
+
+  const checkersName = `${checker?.firstname} ${checker?.surname}`;
+  const checkersEmail = checker?.email || '';
```
[the remark this answers is not in the document]
Comment (abhi-markan) [reply gef-ui/server/controllers/amendments/helpers/map-abandon-amendment-email-variables.ts:39]: If in type `PortalSessionUser` `email` is not optional and since `null` can be removed as a type then I think this can be removed.

**Review by abhi-markan — CHANGES_REQUESTED** [review]

File: gef-ui/server/controllers/amendments/helpers/map-abandon-amendment-email-variables.ts (hunk 4 above)
Comment (abhi-markan) [reply gef-ui/server/controllers/amendments/helpers/map-abandon-amendment-email-variables.ts:29]: Missed this one.

File: gef-ui/server/controllers/amendments/helpers/map-abandon-amendment-email-variables.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  checker,
+}: {
+  deal: Deal | null;
+  facility: Facility | null;
+  amendment: PortalFacilityAmendmentWithUkefId | null;
+  user: PortalSessionUser | null;
+  checker: PortalSessionUser | null;
+}) => {
+  // Check if any of the required parameters are null
+  if (!deal || !facility || !amendment || !user || !checker) {
```
[the remark this answers is not in the document]
Comment (abhi-markan) [reply gef-ui/server/controllers/amendments/helpers/map-abandon-amendment-email-variables.ts:32]: Missed this one.

**Review by abhi-markan — CHANGES_REQUESTED** [review]

File: gef-ui/server/controllers/amendments/cancel-amendment/post-cancel-portal-facility-amendment.ts [hunk 8]
```diff
@@ -18,7 +18,37 @@ export const postCancelPortalFacilityAmendment = async (req: PostCancelPortalFac
...
-    await api.deleteAmendment({ facilityId, amendmentId, userToken });
+    const amendment = await api.getAmendment({ facilityId, amendmentId, userToken });
+
+    if (!amendment) {
+      console.error('Amendment %s was not found for the facility %s', amendmentId, facilityId);
+      return res.redirect('/not-found');
+    }
+
+    if (amendment.status !== PORTAL_AMENDMENT_STATUS.DRAFT) {
+      console.error(`Amendment %s on facility %s is not ${PORTAL_AMENDMENT_STATUS.DRAFT}`, amendmentId, facilityId);
```
Comment (zainUKEF) [inline gef-ui/server/controllers/amendments/cancel-amendment/post-cancel-portal-facility-amendment.ts:29]: not a blocker.  Should ${PORTAL_AMENDMENT_STATUS.DRAFT} be %s? @abhi-markan

**Review by zainUKEF — DISMISSED** [review]

File: portal-api/api-tests/api.js [hunk 9]
```diff
@@ -78,6 +78,10 @@ module.exports = (app) => ({
       get: async (url, query = {}) => request(app).get(url).set(headers).query(query),
 
       remove: async (url) => request(app).delete(url).set(headers).send(),
+
+      removeTo: (data) => ({
```
[the remark this answers is not in the document]
Comment (rkontogianni) [reply portal-api/api-tests/api.js]: done

File: gef-ui/server/controllers/amendments/helpers/map-abandon-amendment-email-variables.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (rkontogianni) [reply gef-ui/server/controllers/amendments/helpers/map-abandon-amendment-email-variables.ts]: done

**Review by abhi-markan — DISMISSED** [review]
LGTM 👍🏼
